joomla-projects/gsoc22_seo
# Open Graph

### Summary of Changes
- Make a separate 'Open Graph' tab under the Article Edit section on the backend
- Provides users the flexibility to modify each article's open graph data, which in turn can be used by social media platforms to display <i>rich objects</i>
- Some of these parameters include
  - Article Type
  - Article Title
  - Article Description
  - Article Image
  - Published Date
  - Article Author
- If any field of the 'Open Graph' tab is left blank, it will use the article's default/original parameters
- Modifications to any of the open graph fields will override the article's default parameters


### Testing Instructions

Perform the following steps for each of the cases defined below
1. Change the parameters as described
2. Save the settings
3. Open the article on the frontend
4. Right click on the article and select 'View page source'
5. Verify the open graph parameters included in the HTML header section

#### CASE 1 - Open Graph tab not used

When all fields in the open graph tab are blank, the article's default parameters are set as its open graph parameters. 
If the article does not use any images, it hasn't been set in the 'Images and Links' tab and it hasn't been set in the OG tab either, the og: image tag does not appear in the screenshot below.

![image](https://user-images.githubusercontent.com/84401192/180204547-f9ae0249-474e-4060-9b7d-d4000566e2a9.png)
![image](https://user-images.githubusercontent.com/84401192/180205432-88959b49-e902-4c27-bb2d-833f0dd27758.png)

#### CASE 2 - Open Graph tab used for all parameters except image

Changing the og parameters under the 'Open Graph' tab will override the default parameters for the article.

![image](https://user-images.githubusercontent.com/84401192/180208472-6cd10a5c-3773-4276-9c0e-1d1057b99232.png)
![image](https://user-images.githubusercontent.com/84401192/180208665-78871dd4-afcc-43ed-b31e-05f309781620.png)


#### CASE 3 - Dealing with multiple images
For the 'og: image' parameter in the HTML header, the following hierarchy is maintained
<ol>
<li>Open Graph Image</li>
<li>Full Article Image</li>
<li>Intro Image</li>
</ol>

For the selected image URL, metadata such as the image type (PNG, JPEG, GIF), height, width don't need to be specified separately by the user. They are extracted automatically from the specified image.

1. When only the article _intro image_ (sample1.png) is set in 'Images and Links' tab, and all other image fields are empty, then the og:image parameter is set to the _intro image_ URL

![image](https://user-images.githubusercontent.com/84401192/180209618-3e47bdc9-6665-44ea-9cdc-7b2bfeeb73f3.png)
![image](https://user-images.githubusercontent.com/84401192/180209809-255eb804-a2a6-4254-b06a-7629c9bf9f85.png)


2. When the article _intro image_ (sample1.png) and the _full article image_ (sample2.png) both are set in 'Images and Links' tab, then the og:image parameter is set to the _full article image_ URL

![image](https://user-images.githubusercontent.com/84401192/180210212-fedb747b-2bea-48a7-b79b-c5947151e259.png)
![image](https://user-images.githubusercontent.com/84401192/180210162-ed418634-78bf-4af0-a5df-b0749a0349fb.png)


3. When the _open graph image_ (sample3.jpg) is set (Open Graph tab) in addition to the above fields, then the og:image parameter is set to the _open graph image_ URL

![image](https://user-images.githubusercontent.com/84401192/180210368-c0b3106f-fe8f-4949-bceb-020e40c00b68.png)
![image](https://user-images.githubusercontent.com/84401192/180210471-6e5da4f5-4034-4c2d-946e-b2570feb1f84.png)
